v3_async_wip/v3_async_wip/exceptions.py

```diff
@@ -0,0 +1,28 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module defines exceptions used by other modules"""
+
+
+# Recoverable
+class RecoverableException(Exception):
+    """An exception that may be recoverable"""
+
+    pass
+
+
+class ConnectionFailedError(RecoverableException):
+    pass
+
+
+# Non-recoverable
+class NonRecoverableException(Exception):
+    """A exception that is not recoverable"""
+
+    pass
+
+
+class AuthorizationError(NonRecoverableException):
```

Comment:
This is tricky, because Auth errors are sometimes recoverable, depending on the type of auth and the level you're operating at. 

For example:
* An x509 connect might fail because the clock is wrong. If some other process fixes the clock and that x509 connect is able to succeed.
* Or the x509 cert has a "not valid until" field, and the cert is invalid until some time passes and then it becomes valid.
* Or an operator might enable (or re-enable) some credentials on the Azure portal and suddenly the client is able to connect. 

You might call thees "external actions", but, if the customer wants a rock-solid device that sits in a closet and never needs maintenence, they should consider an auth error to be recoverable. Even if it tries to reconnect every 30 seconds for 2 months, that might be better (and cheaper) than sending a tech to replace the device.

I suspect the number of "non-recoverable" errors is much smaller than we think. If we define "non-recoverable" as "no matter what happens, this call will never succeed", then a JSON parse error might be non-recoverable. If it fails because the format of the string is wrong it could never succeed. If JSON parse fails because it can't allocate a buffer, it's a little more murky because that parse _could_ succeed if there was more memory.

Likewise, "ConnectionFailed" might happen because the wifi router is unplugged. Try again and it might succeed.

I wonder if "fatal exception" and "non-fatal exception" is a better split, with a very high bar for "fatal", like a fatal error would require a reboot or hardware replacement. I don't know if this makes sense or not, but I think it doesn't make sense.  I don't think "how should the customer respond to this exception" should be part of the exception. Or, if it is part of the exception, it should maybe be a property.

Putting it as a base class makes "Failed connections are never recoverable" part of the API. If we realize that AuthorizationError is actually recoverable, we need to change it's base class, and that's a breaking change.

Reply:
Yeah, I'm not tied to this at all. It's more or less just a placeholder for now, pending future discussions.